linux/Makefile

```diff
@@ -11,15 +11,57 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-TARGETS=all clean
+include engine.mk
+
+ENGINE_LIBRARY=library/libflutter_engine.so
+ENGINE_HEADER=library/include/embedder.h
+TARGETS=all
 SUBDIRS=library example
 
+$(TARGETS): $(SUBDIRS)
+$(SUBDIRS): $(ENGINE_LIBRARY) $(ENGINE_HEADER)
+	$(MAKE) -C $@ $(MAKECMDGOALS)
+
+clean: clean_local
+	rm -f $(ENGINE_LIBRARY)
+	rm -f $(ENGINE_HEADER)
+	rm -f $(FLUTTER_ENGINE_GCLIENT) $(FLUTTER_ENGINE_ROOT_DIR)/.gclient_entries
+	rm -rf $(FLUTTER_ENGINE_SYNC)
+	rm -f $(ENGINE_LIBRARY)
+	rm -f $(ENGINE_HEADER)
+
+# Only cleans this folder (library and example)
+clean_local:
+	make -C library clean
+	make -C example clean
+
 # Example must be built after library. This prevents accidentally building the
 # two in parallel in the event that someone supplies the -j flag to `make`.
 example: library
 
-$(TARGETS): $(SUBDIRS)
-$(SUBDIRS):
-	$(MAKE) -C $@ $(MAKECMDGOALS)
+# Copies library into the linux embedder library/ directory.
+$(ENGINE_LIBRARY): $(FLUTTER_ENGINE_LIB_PATH)
+	cp $(FLUTTER_ENGINE_LIB_PATH) $(dir $(ENGINE_LIBRARY))
+
+# Copies the embedder header into the include/ directory.
+$(ENGINE_HEADER):
+	cp $(FLUTTER_ENGINE_HEADER) $(dir $(ENGINE_HEADER))
+
+$(FLUTTER_ENGINE_LIB_PATH): $(FLUTTER_ENGINE_NINJA_CONFIG)
+
+# Builds libflutter_engine.so
+$(FLUTTER_ENGINE_NINJA_CONFIG): $(FLUTTER_ENGINE_GN_TESTFILE)
+	$(FLUTTER_ENGINE_NINJA) $(FLUTTER_ENGINE_NINJA_ARGS)
+
+# Generates ninja files for building.
+$(FLUTTER_ENGINE_GN_TESTFILE): $(FLUTTER_ENGINE_SYNC)
+	$(FLUTTER_ENGINE_GN) $(FLUTTER_ENGINE_GN_ARGS)
+
+# Runs gclient sync.
```

Comment:
gclient sync is a checkout command; having it run implicitly as part of a build step is not what I would expect. What's the reason for this?

Reply:
Only because it's required to get the correct version of the build tools used to compile the engine (want to make sure not to use the wrong binaries for ninja/gn (they have a specific setup that doesn't seem to generally work with a local version of gn).

Other than that it's just to ensure that the engine can be checked out and built.  This wouldn't happen regularly, e.g. once the user builds the engine once and doesn't change things locally as part of the linux embedder, the only thing that'd be built is the linux embedder.